Packages/com.unity.inputsystem/InputSystem/Plugins/Visualizers/VisualizationHelpers.cs

```diff
@@ -0,0 +1,623 @@
+            where TValue : struct
+        {
+            public ArrayHelpers.RingBuffer<TValue> samples;
+            public ArrayHelpers.RingBuffer<GUIContent> samplesText;
+
+            protected ValueVisualizer(int numSamples = 10)
+            {
+                samples = new ArrayHelpers.RingBuffer<TValue>(numSamples);
+                samplesText = new ArrayHelpers.RingBuffer<GUIContent>(numSamples);
+            }
+
+            public override void AddSample(object value, double time)
+            {
+                var v = default(TValue);
+
+                if (value != null)
+                {
+                    if (!(value is TValue val))
+                        throw new ArgumentException(
+                            $"Expecting value of type '{typeof(TValue).Name}' but value of type '{value?.GetType().Name}' instead",
+                            nameof(value));
+                    v = val;
+                }
+
+                samples.Append(v);
+                samplesText.Append(new GUIContent(v.ToString()));
+            }
+        }
+
+        // Visualizes integer and real type primitives.
+        public class ScalarVisualizer<TValue> : ValueVisualizer<TValue>
+            where TValue : struct
+        {
+            public TValue limitMin;
+            public TValue limitMax;
+            public TValue min;
+            public TValue max;
+
+            public ScalarVisualizer(int numSamples = 10)
+                : base(numSamples)
+            {
+            }
+
+            public override void OnDraw(Rect rect)
+            {
+                // For now, only draw the current value.
+                DrawRectangle(rect, new Color(1, 1, 1, 0.1f));
+                if (samples.count == 0)
+                    return;
+                var sample = samples[samples.count - 1];
+                if (Compare(sample, default) == 0)
+                    return;
+                if (Compare(limitMin, default) != 0)
+                {
+                    // Two-way visualization with positive and negative side.
+                    throw new NotImplementedException();
+                }
+                else
+                {
+                    // One-way visualization with only positive side.
+                    var ratio = Divide(sample, limitMax);
+                    var fillRect = rect;
+                    fillRect.width = rect.width * ratio;
+                    DrawRectangle(fillRect, new Color(0, 1, 0, 0.75f));
+
+                    var valuePos = new Vector2(fillRect.xMax, fillRect.y + fillRect.height / 2);
+                    DrawText(samplesText[samples.count - 1], valuePos, ValueTextStyle);
+                }
+            }
+
+            public override void AddSample(object value, double time)
+            {
+                base.AddSample(value, time);
+
+                if (value != null)
+                {
+                    var val = (TValue)value;
+                    if (Compare(min, val) > 0)
```

Comment:
Do we want to support  the case where min and max can shrink again if values stay within a certain bracket later? I guess depends on what you are visualizing.

Reply:
I think we do. Overall, the plotting functionality here can be improved by leaps and bounds. The current level of "smarts" it shows is pretty poor.